### DISCUSSION TITLE
BIO_get_mem_ptr generates invalid length in openssl 1.1.1w

### DISCUSSION THREAD

**Discussion by timlegge** [openssl/openssl#24791] [Q&A]
Hi

This is on hpux.  I am working on a perl module Crypt::OpenSSL::RSA.  It does the following to extract the data from a BIO:

The printf("length: %li\n", bptr->length); results in:

length: 3827763654656

and the attempt to allocate memory for the data results with newSVpv in:

Out of memory!

```
SV* extractBioString(BIO* p_stringBio)
{
    SV* sv;
    BUF_MEM* bptr = NULL;

    CHECK_OPEN_SSL(BIO_flush(p_stringBio) == 1);
    BIO_get_mem_ptr(p_stringBio, &bptr);

printf("length: %li\n", bptr->length);
    sv = newSVpv(bptr->data, bptr->length);

    CHECK_OPEN_SSL(BIO_set_close(p_stringBio, BIO_CLOSE) == 1);
    BIO_free(p_stringBio);
    return sv;
}
```

Based on some other issues/discussions I replaced it with which does essentially the same thing with :

```
SV* extractBioString(BIO* p_stringBio)
{
    SV* sv;
    char *datap;
    long datasize = 0;

    CHECK_OPEN_SSL(BIO_flush(p_stringBio) == 1);

    datasize = BIO_get_mem_data(p_stringBio, &datap);
    sv = newSVpv(datap, datasize);

    CHECK_OPEN_SSL(BIO_set_close(p_stringBio, BIO_CLOSE) == 1);
    BIO_free(p_stringBio);
    return sv;
}
```
So I guess my question is, is there any reason I should not do that for all versions the module supports?  I realize 1.1.1w is not supported but should I log this as an issue?

Tim

**Comment by nhorman** [comment]
Yes, you should make the change.  Generally speaking those structures are meant to be opaque, and accessing them directly can lead to unpredictable results.
